usr/local/share/bastille/bootstrap.sh

```diff
@@ -441,7 +441,13 @@ case "${1}" in
 NAME_VERIFY=$(echo "${RELEASE}" | grep -iwE '^([1-9]{2,2})\.[0-9](-RELEASE|-RC[1-2])$' | tr '[:lower:]' '[:upper:]')
 if [ -n "${NAME_VERIFY}" ]; then
     RELEASE="${NAME_VERIFY}"
-    UPSTREAM_URL="http://ftp.freebsd.org/pub/FreeBSD/releases/${HW_MACHINE}/${HW_MACHINE_ARCH}/${RELEASE}"
+    if ping -c1 -t2 ftp.freebsd.org >/dev/null 2>&1; then
+        ## main http
+        UPSTREAM_URL="http://ftp.freebsd.org/pub/FreeBSD/releases/${HW_MACHINE}/${HW_MACHINE_ARCH}/${RELEASE}"
+    else
+        ## main ftp
+        UPSTREAM_URL="ftp://ftp.freebsd.org/pub/FreeBSD/releases/${HW_MACHINE}/${HW_MACHINE_ARCH}/${RELEASE}"
+    fi
```

Comment:
Can you tell me the logic behind http/ftp decided by ping?

Reply:
Hi, a very quick ping test will be done on the main HTTP repo and success will return zero and continue, otherwise will switch to the alternate FTP repo, or an alternate mirror in case of HBSD.

This can also be done at the bootstrap/MANIFEST fetch section but the url validate should be preferably done before entering fetch.

Quick return examples of the ping under few simulated conditions:

~~~
# ping -c1 -t2 freebsd.org >/dev/null 2>&1; echo "=> $?"
=> 0
# ping -c1 -t2 freebsdXY.org >/dev/null 2>&1; echo "=> $?"
=> 68
# ping -c1 -t2 freebsd.com >/dev/null 2>&1; echo "=> $?"
=> 2
# ping -c1 -t2 freebsd.org >/dev/null 2>&1; echo "=> $?"
Alarm clock
=> 142
# 
~~~

Time consumed by the quick url validate:
~~~
root@server: ~# time ping -c1 -t2 ftp.freebsd.org
PING ftp.geo.freebsd.org (96.47.72.72): 56 data bytes
64 bytes from 96.47.72.72: icmp_seq=0 ttl=51 time=61.344 ms

--- ftp.geo.freebsd.org ping statistics ---
1 packets transmitted, 1 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 61.344/61.344/61.344/0.000 ms
0.000u 0.003s 0:00.06 0.0%	0+0k 0+0io 0pf+0w
root@server: ~#
~~~

Regards